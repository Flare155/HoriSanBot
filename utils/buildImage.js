const path = require('node:path');
const puppeteer = require("puppeteer");

const buildImage = async (route, data) => {
    const browser = await puppeteer.connect({
        browserURL: 'http://localhost:9222' // Use the same port you specified above
      });

    const page = await browser.newPage();
    page.setViewport({
    width: 1500,
    height: 1000
    })    

    await page.goto(`file:${path.join(__dirname, "hori-visuals", "prod", "index.html")}`);

    await page.evaluate((route, data) => {
        window.path = route;
        window.puppeteerData = data; 
    }, route, data);    

    await page.waitForNetworkIdle();

    const image = await page.screenshot({
    type: "png",
    clip: {
        width: 1200,
        height: 800,
        x : 0,
        y : 0
    }});    
    return image;
}

module.exports = { buildImage };