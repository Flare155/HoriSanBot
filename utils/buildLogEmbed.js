const { footerCreator } = require('./formatting/logFooterCreator.js');
const { calculateEmbedColor } = require('./formatting/calculateEmbedColor.js');
const { EmbedBuilder } = require('discord.js');

// Utility function to create and send the embed message
function buildLogEmbed(interaction, log) {
    // Calculate title and description for embed
    const unit = log.amount.unit;
    const totalSeconds = log.amount.totalSeconds;
    const count = log.amount.count;
    const medium = log.medium;
    const notes = log.notes;
    const coefficient = log.amount.coefficient;

    // Calculate title and description for embed
    const description = unit === "Episodes"
        ? `${Math.round((coefficient * 10) / 60) / 10} minutes/episode ‚Üí +${Math.round((totalSeconds * 10) / 60) / 10} points`
        : `1 point/min ‚Üí +${Math.round((totalSeconds * 10) / 60) / 10} points`;
    let embedTitle = `üéâ ${interaction.member.displayName} logged ${count} ${unit} of ${medium}!`;
    if (unit !== "Episodes") {
        embedTitle = `üéâ ${interaction.member.displayName} logged ${Math.round((totalSeconds * 10) / 60) / 10} minutes of ${medium}!`;
    }


    // Calculate the embed color based on the points
    const embedColor = calculateEmbedColor(totalSeconds);
    // Create footer message
    const footer = footerCreator(interaction, totalSeconds);

    const logEmbed = new EmbedBuilder()
        .setColor(embedColor)
        .setTitle(embedTitle)
        .setDescription(description)
        .setThumbnail(interaction.user.displayAvatarURL())
        .addFields({ name: 'üìñ Title', value: log.title, inline: true })
        .setTimestamp();
    if (notes) {
        logEmbed.addFields({ name: 'üìù Notes', value: log.notes, inline: true });
    }
    logEmbed.setFooter(footer);

    return logEmbed;
}

module.exports = { buildLogEmbed };