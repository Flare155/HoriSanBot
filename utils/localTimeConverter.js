// Takes a date and timezone and oututs the date in the users timezone

const localTimeConverter = (date, timezone) => {
  return new Intl.DateTimeFormat('en-US', {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    timeZone: timezone
  }).format(new Date(date));
};

module.exports = { localTimeConverter};
