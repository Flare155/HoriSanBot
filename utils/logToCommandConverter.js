const { DateTime } = require('luxon');

const logToCommandConverter = (logModel, timezone = 'utc') => {
    const todayUTC = DateTime.utc().toFormat('yyyy-MM-dd');
    const logDateWithTimezone = DateTime.fromJSDate(logModel.timestamp, { zone: timezone }).toFormat('yyyy-MM-dd');

    const isSameDay = todayUTC === logDateWithTimezone;
    let command = isSameDay ? '/log' : `/backlog  date: ${logDateWithTimezone} `;
    const unit = convertUnitToString(logModel.amount.unit);
    command += ` medium: ${logModel.medium} amount: ${logModel.amount.count}${unit} title: ${logModel.title}`;
    if (logModel.notes) {
        command += ` notes: ${logModel.notes}`;
    }
    if (logModel.medium === 'Anime') {
        const unitLength = logModel.amount.unitLength;
        command += ` episode_length: ${unitLength}s`;
    }

    return command;
};

const convertUnitToString = (unit) => {
    if (unit === 'Episodes') {
        return 'ep';
    }

    return 's';
}

module.exports = { logToCommandConverter }