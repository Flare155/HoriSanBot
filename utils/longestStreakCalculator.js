const User = require("../models/User");

const calculateLongestStreak = async (userId, streak) => {
    try {
        // Find user to check if a user exists
        const user = await User.findOne({ userId: userId });

        if (!user) {
            return 0; // If user not found, return 0 streak
        }

        let currectLongestStreak = 0;

        // Checks if the longest streak is smaller then the currect streak
        if(currectLongestStreak < streak) {
            currectLongestStreak = streak;
        }

        return currectLongestStreak;
    } catch (error) {
        console.error("Error calculating streak:", error);
        return 0; // Default to 0 in case of any errors
    }
}

module.exports = { calculateLongestStreak };