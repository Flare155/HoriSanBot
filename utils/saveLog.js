const User = require("../models/User");
const Log = require("../models/Log");

// Function to save the log to the database
async function saveLog(interaction, customDate, medium, title, notes, isBackLog, unit, count, coefficient, totalSeconds) {
    try {
        // Validate required parameters
        if (!interaction || !interaction.user || !interaction.guild) {
            throw new Error("Invalid interaction object.");
        }
        if (!medium || !title || !unit) {
            throw new Error("Missing required parameters.");
        }
    
        // Determine the log date
        const logDate = customDate ? new Date(customDate) : new Date();

        // Construct the amount object
        const amount = {
            unit,
            count,
            totalSeconds,
        };

        // Add coefficient if provided
        if (coefficient) {
            amount.coefficient = coefficient;
        }

        // Save the log entry
        const newLog = new Log({
            userId: interaction.user.id,
            guildId: interaction.guild.id,
            timestamp: logDate,
            medium,
            title,
            notes,
            isBackLog,
            amount,
        });
        await newLog.save();

        // Check if the user exists
        const existingUser = await User.findOne({ userId: interaction.user.id });
        if (!existingUser) {
            // Create a new user if they don't exist
            const newUser = new User({
                userId: interaction.user.id,
                guildId: interaction.guild.id,
                timestamp: new Date().toISOString(),
                streak: 0,
                displayName: interaction.user.displayName,
            });
            await newUser.save();
        } else {
            // Check if the display name has changed
            if (existingUser.displayName !== interaction.user.displayName) {
                existingUser.displayName = interaction.user.displayName;
                await existingUser.save();
            }
        }

        return newLog;
    } catch (error) {
        console.error("Error saving log:", error);
        await interaction.editReply("An error occurred while saving your log.");
    }
}

module.exports = { saveLog };
