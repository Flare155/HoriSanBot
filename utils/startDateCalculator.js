const { DateTime } = require('luxon');

function startDateCalculator(timePeriod, userTimezone = 'UTC') {
  const now = DateTime.now().setZone(userTimezone); // Get the current time in the user's timezone
  let startDate;

  switch (timePeriod) {
    case 'All Time':
      startDate = DateTime.fromMillis(0).setZone(userTimezone); // Unix epoch start
      break;
    case 'Yearly':
      startDate = now.startOf('year'); // Start of the current year
      break;
    case 'Monthly':
      startDate = now.startOf('month'); // Start of the current month
      break;
    case 'Weekly':
      monday = now.startOf('week'); // Start of the current day
      startDate = monday.minus({days: 1})
      break;
    case 'Daily':
      startDate = now.startOf('day'); // Start of the current day
      break;
    default:
      startDate = DateTime.fromMillis(0).setZone(userTimezone); // Default to Unix epoch if unrecognized period
  }

  return startDate.toJSDate(); // Convert to native JS Date object
}

module.exports = { startDateCalculator };
