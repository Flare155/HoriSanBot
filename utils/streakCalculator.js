const { DateTime } = require('luxon');
const Log = require("../models/Log");
const User = require("../models/User");

const calculateStreak = async (userId, guildId) => {
    try {

        // Find the user to get their timezone
        const user = await User.findOne({ userId: userId });
        if (!user) {
            return 0; // If user not found, return 0 streak
        }

        const userTimezone = user.timezone || 'UTC'; // Default to UTC if timezone is not set

        // Find all logs for the user, sorted by timestamp descending
        const logs = await Log.find({ userId: userId }).sort({ timestamp: -1 });

        if (logs.length === 0) {
            return 0; // No logs mean no streak
        }

        let currentStreak = 0;
        const today = DateTime.now().setZone(userTimezone).startOf('day').plus({ hours: 4 }); // "Start of today" is 4 AM in user's timezone
        const yesterday = today.minus({ days: 1 }); // "Start of yesterday" is also based on 4 AM

        // Convert the first log's timestamp to the adjusted "day" (4 AM start)
        let mostRecentLogTime = DateTime.fromJSDate(logs[0].timestamp).setZone(userTimezone).startOf('day').plus({ hours: 4 });

        // Check if the most recent log is from today or yesterday (with the 4 AM reset)
        if (mostRecentLogTime >= yesterday) {
            currentStreak = 1; // Start streak if most recent log is within today or yesterday's window
        } else {
            return 0; // If the most recent log is older than yesterday, reset streak to 0
        }

        // Loop through the remaining logs to check for consecutive "days"
        for (let i = 1; i < logs.length; i++) {
            let currentLogTime = DateTime.fromJSDate(logs[i].timestamp).setZone(userTimezone).startOf('day').plus({ hours: 4 });

            const daysDifference = mostRecentLogTime.diff(currentLogTime, 'days').days;

            if (daysDifference === 1) {
                // Logs are on consecutive days (adjusted for 4 AM window), increment streak
                currentStreak++;
            } else if (daysDifference > 1) {
                // Non-consecutive day found, break the streak
                break;
            }

            mostRecentLogTime = currentLogTime;
        }

        return currentStreak;
    } catch (error) {
        console.error("Error calculating streak:", error);
        return 0; // Default to 0 in case of any errors
    }
};

module.exports = { calculateStreak };
