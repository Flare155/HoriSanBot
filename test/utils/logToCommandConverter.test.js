import {myTest} from '../fixtures';
import { logToCommandConverter } from '../../utils/logToCommandConverter';
import { expect } from 'vitest';

myTest('should convert listening command', ({listeningLog}) => runSnapshotTest(listeningLog));
myTest('should convert watchtime command', ({watchtimeLog}) => runSnapshotTest(watchtimeLog));
myTest('should convert readtime command', ({readtimeLog}) => runSnapshotTest(readtimeLog));
myTest('should convert youtube command', ({youTubeLog}) => runSnapshotTest(youTubeLog));
myTest('should convert manga command', ({mangaLog}) => runSnapshotTest(mangaLog));
myTest('should convert anime command', ({animeLog}) => runSnapshotTest(animeLog));

const runSnapshotTest = (log) => {
    const command = logToCommandConverter(log);
    expect(command).toMatchSnapshot();
}