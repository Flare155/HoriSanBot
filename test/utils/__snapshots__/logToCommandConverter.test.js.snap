// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should convert anime command 1`] = `"/log medium: Anime amount: 2ep title: Anime test notes: Anime note episode_length: 2000s"`;

exports[`should convert listening command 1`] = `"/log medium: Listening amount: 1s title: listening test notes: listening note"`;

exports[`should convert manga command 1`] = `"/log medium: Manga amount: 1s title: Manga test notes: Manga note"`;

exports[`should convert readtime command 1`] = `"/log medium: Readtime amount: 1s title: Readtime test notes: Readtime note"`;

exports[`should convert watchtime command 1`] = `"/log medium: Watchtime amount: 1s title: WatchTime test notes: WatchTime note"`;

exports[`should convert youtube command 1`] = `"/log medium: YouTube amount: 1s title: YouTube test notes: YouTube note"`;
